### PR TITLE
8237467: jlink plugin to save the argument files as input to jlink in the output image

### PR DESCRIPTION
**Pull request by dougxc**
This PR adds a new jlink plugin (`--save-jlink-argfiles=<filenames>`) to support persisting jlink options.

```
> echo "--add-modules jdk.internal.vm.ci --add-options=-Dfoo=xyzzy --vendor-version=\"XyzzyVM 3.14.15\" --vendor-bug-url=https://bugs.xyzzy.com/" > my_image.argfile
> export ALL_MODULES=$(java --list-modules | sed 's:@.*::g' | tr '\n' ',')
> jlink --keep-packaged-modules=my_image/jmods --add-modules $ALL_MODULES --output=my_image --save-jlink-argfiles my_image.argfile
> my_image/bin/java -XshowSettings:properties --version 2>&1 | grep yzzy
> my_image/bin/jlink --add-modules=java.base --output=my_image2
> my_image2/bin/java -XshowSettings:properties --version 2>&1 | grep yzzy
    foo = xyzzy
    java.vendor.url.bug = https://bugs.xyzzy.com/
    java.vendor.version = XyzzyVM 3.14.15
OpenJDK Runtime Environment XyzzyVM 3.14.15 (build 20-internal-2022-09-22-0951036.dnsimon...)
OpenJDK 64-Bit Server VM XyzzyVM 3.14.15 (build 20-internal-2022-09-22-0951036.dnsimon..., mixed mode)
> my_image2/bin/java -d jdk.internal.vm.ci | head -1
jdk.internal.vm.ci@20-internal
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8237467](https://bugs.openjdk.org/browse/JDK-8237467): jlink plugin to save the argument files as input to jlink in the output image
 * [JDK-8294485](https://bugs.openjdk.org/browse/JDK-8294485): jlink plugin to save the argument files as input to jlink in the output image (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [91938caa](https://git.openjdk.org/jdk/pull/10445/files/91938caa8936dea5d897b8f83702ed52b8e7bc4f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10445/head:pull/10445` \
`$ git checkout pull/10445`

Update a local copy of the PR: \
`$ git checkout pull/10445` \
`$ git pull https://git.openjdk.org/jdk pull/10445/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10445`

View PR using the GUI difftool: \
`$ git pr show -t 10445`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10445.diff">https://git.openjdk.org/jdk/pull/10445.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 1]
```diff
@@ -230,6 +231,17 @@ int run(String[] args) {
         }
         Path outputPath = null;
         try {
+            try (InputStream savedOptions = JlinkTask.class.getModule().getResourceAsStream("jdk/tools/jlink/internal/options")) {
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:234]: nit: wrap this long line that helps future side-by-side review.

Comment (mlchung) [reply src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:234]: Nit: define a static variable for this resource file name `jdk/tools/jlink/internal/options` that the plugin can use.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java [hunk 2]
```diff
@@ -230,6 +231,17 @@ int run(String[] args) {
...
+                if (savedOptions != null) {
+                    List<String> newArgs = new ArrayList<>();
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java:236]: `s/newArgs/prependArgs/` to make it clearer.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SaveJlinkArgfilesPlugin.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            return Files.readString(Path.of(argfile));
+        } catch (IOException e) {
+            throw new PluginException("Argfile " + argfile + " is not readable");
+        }
+    }
+
+    @Override
+    public ResourcePool transform(ResourcePool in, ResourcePoolBuilder out) {
+        in.transformAndCopy(Function.identity(), out);
+        byte[] savedOptions = argfiles.stream().
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SaveJlinkArgfilesPlugin.java:96]: Formatting consistent with jlink files:

```suggestion
        byte[] savedOptions = argfiles.stream()
                                      .collect(Collectors.joining("\n"))
                                      .getBytes(StandardCharsets.UTF_8);
```

File: test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            }, module)
+            .assertSuccess();
+        helper.checkImage(image, module, null, null);
+
+        Path launcher = image.resolve("bin/java" + exe);
+        var oa = ProcessTools.executeProcess(launcher.toString(), "-XshowSettings:properties", "--version");
+
+        // Check that the primary image creation ignored the saved args
+        oa.shouldHaveExitValue(0);
+        oa.shouldNotMatch("yzzy");
```
Comment (mlchung) [inline test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java:81]: Can this check for `java.vendor.version = XyzzyVM 3.14.15`, that will also help the reader to understand what the test tries to verify.

File: test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        launcher = image.resolve("bin/jlink" + exe);
+        oa = ProcessTools.executeProcess(launcher.toString(), "--output", image2.toString(), "--add-modules=java.base");
+        oa.shouldHaveExitValue(0);
+
+        // Ensure the saved `--add-options` and `--vendor-*` options
+        // were applied when creating the secondary image.
+        launcher = image2.resolve(Path.of("bin", "java" + exe));
+        oa = ProcessTools.executeProcess(launcher.toString(), "-XshowSettings:properties", "--version");
+        oa.shouldHaveExitValue(0);
+        expectNMatchingLines(oa.getStdout(), "yzzy", 2);
```
Comment (mlchung) [inline test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java:94]: Same for line 94-95: check for `java.vendor.version = XyzzyVM 3.14.15` instead

File: test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java (hunk 4 above)
Comment (dougxc) [reply test/jdk/tools/jlink/plugins/SaveJlinkArgfilesPluginTest.java:81]: Not sure follow as this is a negative match.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SaveJlinkArgfilesPlugin.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public void configure(Map<String, String> config) {
+        var v = config.get(getName());
+
+        if (v == null)
+            throw new AssertionError();
+
+        for (String argfile : v.split(File.pathSeparator)) {
+            argfiles.add(readArgfile(argfile));
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SaveJlinkArgfilesPlugin.java:83]: This plugin should validate if `jdk.jlink` is linked in the resulting image; if not, it should throw an exception.

**Review by mlchung — APPROVED** [review]
Looks good.

**Comment by dougxc** [comment]
Thanks for the reviews @mlchung !

**Pull request closed without merging**